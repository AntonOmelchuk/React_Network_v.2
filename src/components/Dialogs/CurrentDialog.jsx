import React from 'react';
import style from './Dialogs.module.css'
import avatar from '../../assets/avatars/eliot.jpg'
import {connect} from "react-redux";

const CurrentDialog = (props) => {

    if(props.dialog === null) {
        return <h3>No messages</h3>
    } else {

        const {messages, user} = props.dialog;

        return (
            <div className={style.current__dialog}>
                {messages.map(message => (
                    <div className={message.isYour ? style.profile__message : style.message}>
                        <div className={style.current__avatar}>
                            <img src={message.isYour ? avatar : user} alt='avatar' />
                        </div>
                        <div className={style.current__message}>
                            <p>{message.text}</p>
                        </div>
                    </div>
                ))}
            </div>
        );
    }
};

const mapStateToProps = state => ({
    dialog: state.dialogs.current
});

export default connect(mapStateToProps, null)(CurrentDialog);
