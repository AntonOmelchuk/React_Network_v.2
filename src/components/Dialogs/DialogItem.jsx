import React from 'react';
import { A } from 'hookrouter';
import ava from '../../assets/avatars/eliot.jpg';
import style from './Dialogs.module.css';

const DialogItem = ({ dialog, setCurrentDialog }) => {
    const { id, user, lastMessage, isRead, yourLast } = dialog;

    const handleClick = () => {
        setCurrentDialog(dialog);
    };

    return (
        <A className={style.dialog} href={`/dialogs/${id}`} onClick={handleClick}>
            <div className={style.avatar}>
                <img src={user} alt='user avatar' />
            </div>
            <div
                className={
                    isRead
                        ? style.lastMessage__wrapper
                        : style.lastMessage__wrapper + ' ' + style.unread
                }
            >
                <div className={style.lastMessage}>
                    {lastMessage}
                    <div className={style.profile__avatar}>
                        <img src={yourLast ? ava : user} alt='user avatar' />
                    </div>
                </div>
            </div>
        </A>
    );
};

export default DialogItem;
