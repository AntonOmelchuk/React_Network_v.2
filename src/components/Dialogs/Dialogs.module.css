.wrapper {
    width: 750px;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: center;
}

.dialog {
    width: 90%;
    height: 54px;
    margin-top: 12px;
    padding: 9px 12px;

    display: flex;

    background-color: rgba(0, 0, 0, 0.91);
    border-radius: 3px;

    color: antiquewhite;
}

.avatar {
    width: 54px;
    height: 54px;
}

.avatar img {
    width: 100%;
    height: 100%;
    border-radius: 50%;
}

.lastMessage__wrapper {
    width: 80%;
    margin-left: 54px;
    padding: 0 12px;

    background-color: rgba(23, 49, 60, 0.91);
    border-radius: 3px;
}

.lastMessage {
    display: flex;
    align-items: center;
    justify-content: space-between;
}

.unread {
    background-color: rgba(51, 117, 139, 0.88);
    color: #0f2027;
    font-weight: 600;
}

.profile__avatar {
    width: 27px;
    height: 27px;

    margin-left: 9px;
}

.profile__avatar img {
    width: 100%;
    height: 100%;
    border-radius: 50%;
}
