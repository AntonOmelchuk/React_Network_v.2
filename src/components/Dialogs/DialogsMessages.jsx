import React, {useState} from 'react';
import style from './Dialogs.module.css';
import TextareaAutosize from 'react-autosize-textarea';

const DialogsMessages = React.memo(({messages, sendMessage, currentId, deleteMessages}) => {
    const [message, setMessage] = useState('');
    const [deleteMessagesId, setDeleteMessagesId] = useState([]);

    const addDeleteMessagesId = id => {
        if(deleteMessagesId.includes(id)) {
            const newDeleteMessages = deleteMessagesId.filter(mId => mId !== id);

            return setDeleteMessagesId(newDeleteMessages);
        }

        setDeleteMessagesId([...deleteMessagesId, id]);
    };

    const cancelDeleteMessages = () => setDeleteMessagesId([]);

    const onDeleteMessage = () => deleteMessages(deleteMessagesId, currentId);

    const onSendMessage = () => {
        sendMessage(currentId, message);
        setMessage('');
    };

    const onKeyDownHandler = e => {
        if(e.key === 'Enter') {
            e.preventDefault();
            if(message.length > 0) {
                onSendMessage();
            }
        }
    };

    const currentDialogMessages = messages.length ? messages.map(m => (
        <div key={m.id} onClick={() => addDeleteMessagesId(m.id)}
            className={deleteMessagesId.includes(m.id) ? style.deleteMessages + ' ' + style.message : style.message}>
            <div className={style.senderInfo}>
                <div className={style.sender}>{m.senderName}</div>
                <div className={style.time}>{m.addedAt.slice(11, 16)}</div>
            </div>
            <div>{m.body}</div>
        </div>
    )) : <h3>No messages</h3>;

    return (
        <div className={style.messages}>
            <div>{currentDialogMessages}</div>
            <div>
                <TextareaAutosize className={style.sendMessageTextarea} value={message}
                    onChange={e => setMessage(e.target.value)} onKeyDown={e => onKeyDownHandler(e)}
                    maxLength={360} />
                <div className={style.buttonsContainer}>
                    <div>
                        {
                            deleteMessagesId.length > 0 &&
                            <DeleteMessagesButtons cancelDeleteMessages={cancelDeleteMessages}
                                onDeleteMessage={onDeleteMessage} />}
                    </div>
                    <button disabled={!message.length} className={style.dialogButton} onClick={onSendMessage}>
                        Send
                    </button>
                </div>
            </div>
        </div>
    );
});

const DeleteMessagesButtons = ({cancelDeleteMessages, onDeleteMessage}) => (
    <>
        <button onClick={cancelDeleteMessages} className={style.dialogButton + ' ' + style.deleteMessagesButton}>
            cancel
        </button>
        <button onClick={onDeleteMessage} className={style.dialogButton + ' ' + style.deleteMessagesButton}>
            delete
        </button>
    </>
);

export default DialogsMessages;
