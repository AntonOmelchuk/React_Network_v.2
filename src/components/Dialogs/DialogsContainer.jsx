import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import {createStructuredSelector} from 'reselect';

import {selectCurrentId, selectDialogs, selectIsLoading, selectMessages} from '../../selectors/dialogsSelectors';
import {getInitDialogs, getMessages, sendMessage} from '../../actions/dialogsActions';

import Dialogs from './Dialogs';

const DialogsContainer = ({
    getInitDialogs,
    getMessages,
    dialogs,
    messages,
    sendMessage,
    isLoading,
    currentId,
    id
}) => {
    useEffect(() => {
        getInitDialogs(id);
        document.title = 'Dialogs';
    }, [getInitDialogs, id]);

    return (
        <Dialogs dialogs={dialogs} isLoading={isLoading} currentId={currentId} sendMessage={sendMessage}
            messages={messages} getMessages={getMessages} />
    );
};

const mapStateToProps = createStructuredSelector({
    dialogs: selectDialogs,
    messages: selectMessages,
    isLoading: selectIsLoading,
    currentId: selectCurrentId
});

export default connect(mapStateToProps, {sendMessage, getInitDialogs, getMessages})(DialogsContainer);
