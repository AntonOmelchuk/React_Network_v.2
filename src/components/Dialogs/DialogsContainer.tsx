import React, {useEffect} from 'react';
import {connect} from 'react-redux';
import {createStructuredSelector} from 'reselect';
import Dialogs from './Dialogs';

import {selectCurrentId, selectDialogs, selectIsLoading, selectMessages} from '../../selectors/dialogsSelectors';
import {deleteMessages, getInitDialogs, getMessages, sendMessage} from '../../actions/dialogsActions';

type OwnPropsType = {
    id: number
}

type MapStatePropsType = {
    dialogs: [],
    messages: [],
    isLoading: boolean,
    currentId: number
}

type MapDispatchPropsType = {
    sendMessage: (userId: number, message: string, fromModal:boolean) => void,
    getInitDialogs: (id: number) => void,
    getMessages: (userId: number) => void,
    deleteMessages: (messages: Array<number>, userId: number) => void
}

type PropsType = OwnPropsType & MapStatePropsType & MapDispatchPropsType

const DialogsContainer: React.FC<PropsType> = ({
    getInitDialogs,
    getMessages,
    deleteMessages,
    dialogs,
    messages,
    sendMessage,
    isLoading,
    currentId,
    id
}) => {
    useEffect(() => {
        getInitDialogs(id);
        document.title = 'Dialogs';
    }, [getInitDialogs, id]);


    return (
        <Dialogs dialogs={dialogs}
                 // @ts-ignore
                 isLoading={isLoading}
                 currentId={currentId}
                 sendMessage={sendMessage} messages={messages}
                 getMessages={getMessages} deleteMessages={deleteMessages} />
    );
};

const mapStateToProps = createStructuredSelector({
    dialogs: selectDialogs,
    messages: selectMessages,
    isLoading: selectIsLoading,
    currentId: selectCurrentId
});

export default connect(mapStateToProps, {sendMessage, getInitDialogs, getMessages, deleteMessages})(DialogsContainer);
