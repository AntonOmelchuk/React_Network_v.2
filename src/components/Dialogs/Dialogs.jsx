import React from 'react'
import DialogItem from './DialogItem'
import {connect} from 'react-redux'
import style from './Dialogs.module.css'
import {setCurrentDialog} from '../../actions/dialogsActions';

const Dialogs = (props) => {

    const chooseDialog = (id) => {
        props.history.push(`/dialogs/${id}`)
    };

    return (
        <div className={style.wrapper}>
            {props.dialogs.map(dialog => (
                <DialogItem key={dialog.id} dialog={dialog} chooseDialog={chooseDialog}
                    setCurrentDialog={props.setCurrentDialog}/>))}
        </div>
    )
};

const mapStateToProps = state => ({
    dialogs: state.dialogs.dialogs
});

export default connect(mapStateToProps, {setCurrentDialog})(Dialogs);
