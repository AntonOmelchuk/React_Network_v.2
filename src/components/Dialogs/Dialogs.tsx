import React from 'react';

import style from './Dialogs.module.css';

import DialogsUsers from './DialogsUsers';
import DialogsMessages from './DialogsMessages';
import {DialogType, MessageType} from '../../../types'

type PropsType = {
  dialogs: Array<DialogType>,
  messages: Array<MessageType>,
  currentId: number,
  sendMessage: (userId: number, message: string, fromModal?:boolean) => void,
  getMessages: (userId: number) => void,
  deleteMessages: (messages: Array<number>, userId: number) => void
}

class Dialogs extends React.Component<PropsType> {
  render() {
    let {dialogs, currentId, sendMessage, getMessages, messages, deleteMessages} = this.props

    if (!dialogs) return 'No dialogs'
    return (
      <div className={style.container}>
        <DialogsUsers dialogs={dialogs} currentId={currentId} getMessages={getMessages}/>
        <DialogsMessages messages={messages} sendMessage={sendMessage} currentId={currentId}
                         deleteMessages={deleteMessages}/>
      </div>
    )
  }
}

export default Dialogs;
