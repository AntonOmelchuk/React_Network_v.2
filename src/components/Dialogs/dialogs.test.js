import {shallow} from 'enzyme';
import React from 'react';

import DialogItem from './DialogItem';
import {CurrentDialog} from './CurrentDialog';
import {Dialogs} from './Dialogs';

describe('DialogItem component', () => {
    let wrapper;
    let mockSetCurrentDialog;

    beforeEach(() => {

        mockSetCurrentDialog = jest.fn();

        const mockProps = {
            dialog: {
                id: 3,
                user: 'test user',
                lastMessage: 'test message',
                isRead: true,
                yourLast: false
            },
            setCurrentDialog: mockSetCurrentDialog
        };

        wrapper = shallow(<DialogItem {...mockProps} />)
    });

    it('renders DialogItem component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('calls setCurrentDialog on click', () => {
        wrapper.find('A').simulate('click');
        expect(mockSetCurrentDialog).toHaveBeenCalled();
    })
});

describe('CurrentDialog component' ,() => {
    let wrapper;
    let mockSendMessage;

    beforeEach(() => {
        mockSendMessage = jest.fn();

        const mockProps = {
            dialog: {
                messages: [
                    {id: 3, isYour: true, text: 'test text'}
                ],
                user: 'test user'
            },
            sendMessage: mockSendMessage
        };

        wrapper = shallow(<CurrentDialog {...mockProps} />);
    });

    it('renders CurrentDialog component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('calls sendMessage on click', () => {
        wrapper.find('button').simulate('click');
        expect(mockSendMessage).toHaveBeenCalled();
    })
});

describe('Dialogs component', () => {
    let wrapper;

    beforeEach(() => {
        const mockProps = {
            dialogs: [],
            setCurrentDialog: jest.fn(),
            isAuth: true
        };

        wrapper = shallow(<Dialogs {...mockProps} />);
    });

    it('renders Dialogs component', () => {
        expect(wrapper).toMatchSnapshot();
    })
});