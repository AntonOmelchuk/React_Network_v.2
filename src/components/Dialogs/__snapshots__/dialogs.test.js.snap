// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentDialog component renders CurrentDialog component 1`] = `ShallowWrapper {}`;

exports[`DialogItem component renders DialogItem component 1`] = `ShallowWrapper {}`;

exports[`Dialogs component renders Dialogs component 1`] = `ShallowWrapper {}`;
